[addon] support to set binary add-on instance settings by python

AlwinEsch opened the pull request:
## Description

This allow a python script to edit add-on instance settings where used within PVR add-ons.

To prevent conflicts by user set within Kodi's GUI are the by python created ones starting with id >1000. 
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?
By a small python script has it works for tests.

```py
import xbmcaddon
import xbmcgui
 
addon       = xbmcaddon.Addon('pvr.demo')
addonname   = addon.getAddonInfo('name')

newid = addon.getFreeNewInstanceId()
addonsettingnew = addon.getSettings(newid)
addonsettingnew.setString('host', '10.144.12.13')
addon.setInstanceIdState(newid, True, 'By Python added')

# addon.deleteInstanceId(1001);

printstring = "Support Instances: " + str(addon.supportsInstanceSettings()) + "\n" \
              "Used Instances:\n"
for x in addon.getKnownInstanceIds():
    printstring += "ID: " + str(x) + "\n"
    printstring += "Test setting: " + addon.getSettings(x).getString('host') + "\n"

xbmcgui.Dialog().textviewer(addonname, printstring)
```

If the instance already present is it enough to call:
```py
id = 1001
instancesetting = addon.getSettings(id)
instancesetting.setString('host', '10.144.12.13')
```

By this the add-on becomes the `ADDON_STATUS SetInstanceSetting(const std::string& settingName, const kodi::addon::CSettingValue& settingValue)` call.

<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

phunkyfish commented on xbmc/interfaces/legacy/Addon.cpp:
```diff
@@ -81,9 +81,61 @@ namespace XBMCAddon
+          throw XBMCAddon::WrongTypeException("Invalid setting type");
+
+        if (value)
+          CServiceBroker::GetAddonMgr().PublishInstanceAdded(pAddon->ID(), instance);
```
Will this callback to the PVR add-on? It is important that it does as otherwise the add-on cannot handle settings changes. Ideally each changed setting would callback to `SetInstanceSetting()` for the add-on implementing the interface kodi::addon::IAddonInstance.

AlwinEsch replied:
> Will this callback to the PVR add-on? It is important that it does as otherwise the add-on cannot handle settings changes. Ideally each changed setting would callback to `SetInstanceSetting()` for the add-on implementing the interface kodi::addon::IAddonInstance.
> 
> IT's also possible this happens for free as part of the regular settings interface.

The callback mainly works to the CPVRClients, where then call the PVR add-ons and only needed on add or disable/enable. `SetInstanceSetting()` now becomes called if instance already know on add-on.

matthuisman replied:
@phunkyfish 
i dont have the env setup to create builds sorry. If you can get me an x86_64 windows I can test.
We want to make sure that we can force IPTV Simple to reload its playlists right?

AlwinEsch replied:
> @phunkyfish i dont have the env setup to create builds sorry. If you can get me an x86_64 windows I can test. We want to make sure that we can force IPTV Simple to reload its playlists right?

This mainly to change the use state about the instance, e.g. the disable/enable. Normally should the add-on handle setting changes by self, if he see a change within in his settings. But optionally can the python also do a hacky disable and enable again to force a reload.

matthuisman commented:
This looks like it should solve https://github.com/xbmc/xbmc/issues/22779

AlwinEsch commented:
> This looks like it should solve #22779

Yes this should the issue fix.

@phunkyfish have updated, should be now a bit better.

The `kodi_addon_instance_enabled` and `kodi_addon_instance_name` becomes no more set direct on addon and use the function `bool Addon::setInstanceIdState(unsigned int instance, bool enabled, const String& name/* = emptyString*/)` 

The `setInstanceIdState` must be then called if the instance is added new or enabled/disabled, without becomes more hard to make the works to PVRManager to set it.

Also found here a small bug as the instance id not given to it: 
https://github.com/xbmc/xbmc/blob/f5ab9fd5d2d6d3589470208a774bbfa757a63ee9/xbmc/addons/settings/AddonSettings.cpp#L442

enen92 commented on xbmc/addons/IAddon.h:
```diff
@@ -46,6 +46,11 @@ constexpr AddonInstanceId ADDON_SINGLETON_INSTANCE_ID = 0;
 constexpr AddonInstanceId ADDON_FIRST_INSTANCE_ID = 1;
 
+/*!
+ * @brief Identifier denoting initial first add-on instance set by an Python script.
```
```suggestion
 * @brief Identifier denoting initial first add-on instance set by a Python script.
```

enen92 commented on xbmc/interfaces/legacy/Addon.h:
```diff
@@ -97,12 +97,158 @@ namespace XBMCAddon
+      ///
+      /// \ingroup python_xbmcaddon
+      /// @brief \python_func{ xbmcaddon.Addon([id]).supportsInstanceSettings() }
+      /// Returns true if asked add-on supports instances
```
```suggestion
      /// Returns true if the requested add-on supports several instances
```

An add-on with a single instance still supports instances (but just one :) )

enen92 commented on xbmc/interfaces/legacy/Addon.h:
```diff
@@ -97,12 +97,158 @@ namespace XBMCAddon
+      /// @brief \python_func{ xbmcaddon.Addon([id]).supportsInstanceSettings() }
+      /// Returns true if asked add-on supports instances
+      ///
+      /// @return                        Returns as a boolean, true if supported, false if not
```
```suggestion
      /// @return                        Returns true if supported, false otherwise
```

enen92 commented on xbmc/interfaces/legacy/Addon.h:
```diff
@@ -97,12 +97,158 @@ namespace XBMCAddon
+      ///
+      /// \ingroup python_xbmcaddon
+      /// @brief \python_func{ xbmcaddon.Addon([id]).getKnownInstanceIds() }
+      /// Returns list about known available instance id's
```
```suggestion
      /// Returns the list of known available instance ids
```

enen92 commented on xbmc/interfaces/legacy/Addon.h:
```diff
@@ -97,12 +97,158 @@ namespace XBMCAddon
+      /// @brief \python_func{ xbmcaddon.Addon([id]).getFreeNewInstanceId() }
+      /// Get free add-on instance id to set new independent instance and settings about.
+      ///
+      /// @note After a set of new instance settings needs an @ref publishInstanceChange() called.
```
```suggestion
      /// @note After setting the new instance settings a @ref publishInstanceChange() call is required.
```

enen92 commented on xbmc/interfaces/legacy/Addon.h:
```diff
@@ -97,12 +97,158 @@ namespace XBMCAddon
+      ///
+      /// @param instance            unsigned integer - Instance identifier to use.
+      /// @param enabled             boolean - To enable or disable the instance
+      /// @param name                [opt] string or unicode - The instance name used inside Kodi's
```
```suggestion
      /// @param name                [opt] string or unicode - The instance name used inside Kodi
```

enen92 replied:
Missing from the signature above?

enen92 commented on xbmc/interfaces/legacy/Addon.h:
```diff
@@ -97,12 +97,158 @@ namespace XBMCAddon
+      /// @param enabled             boolean - To enable or disable the instance
+      /// @param name                [opt] string or unicode - The instance name used inside Kodi's
+      ///                            @warning Needs on new added instances set.
+      /// @return                    True if the set of the instance change was OK, false otherwise
```
```suggestion
      /// @return                    True if the setting the instance change was successful, false otherwise
```

enen92 commented on xbmc/interfaces/legacy/Addon.h:
```diff
@@ -97,12 +97,158 @@ namespace XBMCAddon
+      ///                            @warning Needs on new added instances set.
+      /// @return                    True if the set of the instance change was OK, false otherwise
+      ///
+      /// @note About add or Enable/Disable about an instance this function needs called.
```
```suggestion
      /// @note To add or Enable/Disable an instance this function needs to be called.
```

enen92 commented on xbmc/interfaces/legacy/Addon.h:
```diff
@@ -97,12 +97,158 @@ namespace XBMCAddon
       /// @brief \python_func{ xbmcaddon.Addon([id]).getSettings() }
       /// Returns a wrapper around the addon's settings.
       ///
+      /// @param id                      unsigned integer - by use id of the wanted add-on instance setting
```
```suggestion
      /// @param id                      unsigned integer - used id of the requested add-on instance setting
```

enen92 commented on xbmc/interfaces/legacy/Addon.h:
```diff
@@ -97,12 +97,158 @@ namespace XBMCAddon
       /// Returns a wrapper around the addon's settings.
       ///
+      /// @param id                      unsigned integer - by use id of the wanted add-on instance setting
+      ///                                to access, as default by 0 to normal settings.xml.
```
```suggestion
      ///  @note use 0 to access regular settings.xml.
```

enen92 commented on xbmc/interfaces/legacy/Addon.h:
```diff
@@ -97,12 +97,158 @@ namespace XBMCAddon
+#ifdef DOXYGEN_SHOULD_USE_THIS
+      ///
+      /// \ingroup python_xbmcaddon
+      /// @brief \python_func{ xbmcaddon.Addon([id]).setInstanceIdState() }
```
Missing argument?

enen92 commented on xbmc/interfaces/legacy/Addon.h:
```diff
@@ -97,12 +97,158 @@ namespace XBMCAddon
+      ///
+      /// \ingroup python_xbmcaddon
+      /// @brief \python_func{ xbmcaddon.Addon([id]).setInstanceIdState() }
+      /// Inform about changed on add-on instance system.
```
```suggestion
      /// Set changed state on add-on instance system.
```

AlwinEsch commented:
Thanks a lot @enen92! Have updated the text, also removed the spaces on the value text, added they some years ago, but makes not really sense to use.

AlwinEsch commented:
Have added now also a second commit where allow it within old setting functions, not sure to make as they are set as deprecated.

Further can the python add-on open the instance settings dialog direct.

AlwinEsch commented on xbmc/interfaces/legacy/Addon.cpp:
```diff
@@ -155,129 +158,131 @@ namespace XBMCAddon
-    String Addon::getSetting(const char* id)
+    String Addon::getSetting(const char* id, unsigned int instance /* = 0*/)
     {
       return pAddon->GetSetting(id);
```
NOTE: Have forgot give of "instance" here, but not updated now, as the support there maybe becomes removed again.

emveepee commented:
> The `kodi_addon_instance_enabled` and `kodi_addon_instance_name` becomes no more set direct on addon and use the function `bool Addon::setInstanceIdState(unsigned int instance, bool enabled, const String& name/* = emptyString*/)`
> 

Will this remove the confusing debug message when the  resource/instance-setting.xml is read here since it shouldn't have these values. https://github.com/xbmc/xbmc/blob/8de133330cc47e1d4809f271118b8ac2fb3701eb/xbmc/addons/settings/AddonSettings.cpp#L226C55-L226C55

```
2023-08-22 20:55:54.035 T:22324   debug <CSettingsManager>: requested setting (kodi_addon_instance_name) was not found.
2023-08-22 20:55:54.036 T:22324   debug <CSettingsManager>: requested setting (kodi_addon_instance_enabled) was not found.
```

AlwinEsch commented:
> > The `kodi_addon_instance_enabled` and `kodi_addon_instance_name` becomes no more set direct on addon and use the function `bool Addon::setInstanceIdState(unsigned int instance, bool enabled, const String& name/* = emptyString*/)`
> 
> Will this remove the confusing debug message when the resource/instance-setting.xml is read here since it shouldn't have these values. https://github.com/xbmc/xbmc/blob/8de133330cc47e1d4809f271118b8ac2fb3701eb/xbmc/addons/settings/AddonSettings.cpp#L226C55-L226C55
> 
> ```
> 2023-08-22 20:55:54.035 T:22324   debug <CSettingsManager>: requested setting (kodi_addon_instance_name) was not found.
> 2023-08-22 20:55:54.036 T:22324   debug <CSettingsManager>: requested setting (kodi_addon_instance_enabled) was not found.
> ```

Ah, thanks. Yes this comes on Debug some times and not needed.
I change the Settings function to this and to ignore on some cases:
```cpp
  /*!
   \brief Gets the setting with the given identifier.

   \param id Setting identifier
   \param[in] notPresentPossible If is set, error messages are ignored because the given setting
                                 might not actually exist.
                                 Can refer to background settings related to Kodi, which may be.
                                 added later, e.g. to an add-on.
   \return Setting object with the given identifier or nullptr if the identifier is unknown
   */
  std::shared_ptr<CSetting> GetSetting(const std::string &id, bool notPresentPossible = false) const;
```

emveepee commented:
> ```
>   std::shared_ptr<CSetting> GetSetting(const std::string &id, bool notPresentPossible = false) const;
> ```

Thanks this has bugged me since I first saw them.    Perhaps in a future PR this could lead new to new settings type for removed variables, instead of having to hide them.

AlwinEsch committed:
[addons] fix instance setting save in "CAddonSettings::Save()"

Before was the instance id not given on "m_addon->SaveSettings(...)" call.

AlwinEsch commented:
Updated it, should work now, had recently thrown a bug where name and enable/disable were not saved.

I have removed my changes in the old python API because they are deprecated.

Setting the settings is now also possible in the Python addon class directly via `CAddonSettings` and no longer via the add-on class (because this way is also used in the new python settings API).

From my side it should be ready, which is why I removed the WIP. Take a look and say what you think.

phunkyfish commented on xbmc/settings/SettingsBase.h:
```diff
@@ -103,9 +103,14 @@ class CSettingsBase
    */
-  std::shared_ptr<CSetting> GetSetting(const std::string& id) const;
+  std::shared_ptr<CSetting> GetSetting(const std::string& id,
+                                       bool notPresentPossible = false) const;
```
Suggest renaming `notPresentPossible` to `presenceValidated`, and switch the default to true.

AlwinEsch replied:
Thanks @phunkyfish, have changed. This now much better as name :+1: .

phunkyfish commented on xbmc/settings/lib/SettingsManager.cpp:
```diff
@@ -532,7 +533,9 @@ SettingPtr CSettingsManager::GetSetting(const std::string &id) const
   }
 
-  m_logger->debug("requested setting ({}) was not found.", id);
+  if (!notPresentPossible)
```
And remove the `!` here. /double negatives are bad,

phunkyfish commented on xbmc/addons/settings/AddonSettings.cpp:
```diff
@@ -223,8 +223,8 @@ void CAddonSettings::OnSettingAction(const std::shared_ptr<const CSetting>& sett
 {
-  if (GetSetting(ADDON_SETTING_INSTANCE_NAME_VALUE) ||
-      GetSetting(ADDON_SETTING_INSTANCE_ENABLED_VALUE))
+  if (GetSetting(ADDON_SETTING_INSTANCE_NAME_VALUE, true) ||
+      GetSetting(ADDON_SETTING_INSTANCE_ENABLED_VALUE, true))
   {
```
Change true to false.

AlwinEsch commented:
Specially to the Python team, OK to bring in as this?

phunkyfish commented:
@enen92 could you take a look from the Python side?

matthuisman commented:
Will these new methods work for addons using single settings? Be nice if we
can use the same methods regardless of single/multiple instances.

On Thu, 7 Sept 2023, 7:57 am phunkyfish, ***@***.***> wrote:

> @enen92 <https://github.com/enen92> could you take a look from the Python
> side?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/xbmc/xbmc/pull/23648#issuecomment-1709013910>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABPQAKLCW45JLM4OHZSYVBTXZDIS3ANCNFSM6AAAAAA3Y2YWYA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

enen92 commented on xbmc/addons/Addon.h:
```diff
@@ -361,6 +361,24 @@ class CAddon : public IAddon
+   *                              added later, e.g. to an add-on.
+   * \return true if the setting's value was retrieved, false otherwise.
+   */
+  bool GetKodiInstanceSetting(AddonInstanceId id,
```
We should avoid the usage of the Kodi naming on a public api

enen92 commented on xbmc/settings/SettingsBase.h:
```diff
@@ -103,9 +103,13 @@ class CSettingsBase
    \brief Gets the setting with the given identifier.
 
    \param id Setting identifier
+   \param[in] presenceValidated If is set to false, error messages are ignored because the given
```
Why is this needed? shouldn't it just log the error if the setting does not exist?

emveepee replied:
I suggested this change.  There are two instances settings that shouldn't be in the source instance-settings.xml  that were logged as debug message.  The error is when they are included.

enen92 replied:
Can you provide a more detailed example or explain it better? I am not really aware of the instance stuff. Regardless, the bool is pretty much a bad smell

emveepee replied:
I did earlier https://github.com/xbmc/xbmc/pull/23648#issuecomment-1690799635 These two values are not supposed to be in instance-setting.xml as they are added by core.   The bool is basically saying don't log a validation message

enen92 replied:
Then I'd say the proper way of handling this is to avoid logging in case they are built-in instance settings (the case of those two settings). Create some set with the known ignored values and check against it before logging.

enen92 commented on xbmc/interfaces/legacy/Addon.h:
```diff
@@ -100,10 +100,157 @@ namespace XBMCAddon
+      /// settings.setInt('m3uPathType', 0)
+      /// settings.setString('m3uPath', '/SOME_PATH_OF_YOU')
+      /// ...
+      /// addon.setInstanceIdState(newinstanceid, True, 'Example by Python added')
```
Is this string meaningful for any actual feature or is it just some optional string?
I feel it is just used for some logging or so. If that's the case I'd suggest to drop it. It should be the add-on in charge of logging stuff

AlwinEsch replied:
See above this:
```.cpp
      /// @param name [opt] string or unicode - The instance name used inside Kodi
      ///             @warning Needs on new added instances set.
```

enen92 replied:
Got it by the screenshots you posted. Maybe rename the example as "Example instance created from python" (reads a bit better)

enen92 replied:
Also `@warning Needs on new added instances set.` -> "Required when creating a new settings instance" ?

enen92 commented on xbmc/interfaces/legacy/Addon.h:
```diff
@@ -100,10 +100,157 @@ namespace XBMCAddon
+      /// ..
+      /// ~~~~~~~~~~~~~
+      ///
+      getKnownInstanceIds();
```
I think known is pretty much implicit. Maybe rename to getInstanceIds()?

AlwinEsch replied:
Have changed it to `getInstanceIds`, outside of python I leaved still as `GetInstanceIds()`, maybe I change in another request to have all with same name a again.

enen92 commented on xbmc/addons/Addon.cpp:
```diff
@@ -462,6 +462,33 @@ bool CAddon::GetSettingString(const std::string& key,
+bool CAddon::GetKodiInstanceSetting(AddonInstanceId id,
+                                    bool& enabled,
+                                    std::string& name,
+                                    bool presenceValidated)
```
Is the presenceValidated mandatory? Why not return false if the setting does not exist?
I'd say it makes the api a bit cumbersome for no reason

AlwinEsch replied:
I can remove as value, thought only for cases if in future this becomes called on place where needs the presence of settings.

enen92 commented on xbmc/interfaces/legacy/Addon.h:
```diff
@@ -100,10 +100,157 @@ namespace XBMCAddon
+      /// ~~~~~~~~~~~~~
+      ///
+      getFreeNewInstanceId();
+#else
```
Same here. Why not simply `XX = addon.InstanceId()`?
Cleaner

AlwinEsch replied:
Thats as `addon.instance()` I see really confusing, as I expect then a handling class and not an identifier as return.
Maybe another, but has not found a good name for it.

enen92 commented on xbmc/interfaces/legacy/Addon.h:
```diff
@@ -100,10 +100,157 @@ namespace XBMCAddon
+      ///
+      setInstanceIdState(...);
+#else
+      bool setInstanceIdState(unsigned int instance,
```
It seems to me that this function sets the state of an instance, not really of the Id. So preferably: `setInstanceState(Id, value)`

AlwinEsch replied:
Yes, as `setInstanceState(Id, value)` is better and becomes changed.

enen92 commented on xbmc/interfaces/legacy/Addon.h:
```diff
@@ -100,10 +100,157 @@ namespace XBMCAddon
+      /// ..
+      /// ~~~~~~~~~~~~~
+      ///
+      deleteInstanceId(...);
```
deleteInstance(Id)?

AlwinEsch replied:
Is better, becomes changed.

phunkyfish commented:
> Will these new methods work for addons using single settings? Be nice if we can use the same methods regardless of single/multiple instances.
> […](#)

Can you give an example of how this would work?

matthuisman commented:
getKnownInstanceIds returns a single [ID] that represents the single settings

AlwinEsch committed:
[addons] support to set binary add-on instance settings by python

This allow a python script to edit add-on instance settings where used within PVR add-ons.

To prevent conflicts by user set within Kodi's GUI are the by python created ones starting with id >1000.
By a small python script has it works for tests.

CastagnaIT commented:
i cant review because from first implementation in kodi of "multiple instances binary addons" i never understand his purpose and how works and if it works for PVR only or all binary, then with the scripts in the description its not clear to me what are you doing..
would be better having a wiki page (or exists already?) that explain what this is this and how works / how implement and/or a sample project

AlwinEsch committed:
[addons] avoid log calls that setting doesn't exist and isn't expected either

Previously, there were always calls to Kodi's own add-on instance settings in the debug log, since these cannot be known anyway, it makes no sense to set a log message.

AlwinEsch commented:
> i cant review because from first implementation in kodi of "multiple instances binary addons" i never understand his purpose and how works and if it works for PVR only or all binary, then with the scripts in the description its not clear to me what are you doing.. would be better having a wiki page (or exists already?) that explain what this is this and how works and for what uses / how implement and/or a sample project

The instances really only refer to binary add-ons and currently only used in PVR. The background to Python is that an add-on may be able to change the settings of a PVR add-on. E.g. in pvr.iptvsimple where a python script can then set everything suitable for multiple PVR stream backend servers by adding these necessary instances with the associated settings to the binary add-on.

AlwinEsch committed:
[addons] allow direct instance selection in add-on settings dialog

This makes it possible to call up an instance setting immediately without first showing the selection dialog.

CastagnaIT commented:
yes i more or less understand that script above change some settings of an "instance" (you create always a new one?)
but i still don't have a complete background on how "multiple instances binary addons" works,
as for example, who/where create these addon instances, how run them, they are indipendent like a service? and then has nothing to do with media playback? so a wiki would be better, an external user dont understand nothing, unlikely that anyone would use this feature

AlwinEsch commented:
> getKnownInstanceIds returns a single [ID] that represents the single settings So could be used for an old or new version of an addon with/without multiple instances.
> 
> This may already be the case :)
> 
> Not a huge deal to be honest. just makes the calling code a bit easier
> 
> Im happy enough to do:
> 
> ```
> ids = addon.getKnownInstanceIds()
> settings = addon.getSettings(id[0] if ids else None)
> ```

Hi @matthuisman is not directly so easy, a add-on can still have both, within id 0 are general settings of add-on where relates to all instances on it. All numbers higher 0 then relates to instance only.

The `settings = addon.getSettings(id[0] if ids else None)` can be used, only then limited to first available instance and can not directly identify if it is by user set or from python script. Also then works like before, as there no instances supported.


> yes i more or less understand that script above change some settings of an "instance" (you create always a new one?) but i still don't have a complete background on how "multiple instances binary addons" works, as for example, who/where create these addon instances, how run/stop them, they are indipendent like a service? and then has nothing to do with media playback? so a wiki would be better, an external user dont understand nothing, unlikely that anyone would use this feature

Hello @CastagnaIT, yes you right, there should be some documentations about to know what exactly is the background about. **Will look in next time what would be the best about and to add documentations.**
About "(you create always a new one?), not directly. Normally the user add in the settings dialog one or more instances and edit they to match related backends. This are then selectable and editable within a dialog.

![Screenshot_20230907_085330](https://github.com/xbmc/xbmc/assets/6879739/63f20c4a-4ba5-4afd-9e72-cdcfe2c81c23)

The "Edit add-on settings" there relates as button to global general settings of PVR add-on (only visable if add-on have a standard "settings.xml", they below are the independent instance settings on the add-on, the last added by the python script.
The both buttons above are for the user to add or remove a instance.

Python can then with his new API then add, delete or edit all instances, they from user or they added from him in script.

About "they are indipendent like a service?" not directly, the add-on is like a app, it runs all instances independent from each other in own processes, but everything in same base and can if needed also interact between.

enen92 commented on xbmc/interfaces/legacy/Addon.h:
```diff
@@ -100,10 +100,155 @@ namespace XBMCAddon
+#ifdef DOXYGEN_SHOULD_USE_THIS
+      ///
+      /// \ingroup python_xbmcaddon
+      /// @brief \python_func{ xbmcaddon.Addon([id]).setInstanceState(id, enabled, [name]) }
```
Why is the name set when change the state on the instance and not right when the instance is created? Can it be moved to getFreeNewInstanceId()?

AlwinEsch replied:
There is a big hen and egg problem, by `getFreeNewInstanceId()` is and can the instance not created, as the needed settings for it not available and PVR system then start works where always ends in errors, thats why it only give a free id. 

The fields about "enabled" and "name" are parts only handled by Kodi itself, about "enabled" is easy to understand why, about name relates to the name used on the GUI in Kodi (see picture above).

The `setInstanceState` needs be called on new added instances, thats why name needed on it, or can be called too on already present ones to change name and/or enable / disable it, thats why name declared optional.

Maybe the function can become another name. Have you a idea about?

----------------------

Another solution can be, but then it becomes hacky and needs more functions and much changes to add:
| Function | Hacky description |
|--|--|
| `settings = addon.createInstanceSettings(enabled, name)` | It returns the settings class and not an id<br> - The enabled and name needs temporary stored somewhere<br> - access to `CSettingsBase` in Kodi no more enough and needs access to `CAddonSettings` as this handle the id.
| `settings.setAddedInstanceReady()`  or<br> `addon.setAddedInstanceReady(settings)`| Something needs always done to report Kodi that new instance is ready to start and thats PVR system does it<br><br>Maybe it can optionally also be done by the settings class destructor, but then Python needs to do destruct it and never stores it somewhere.
| `id = settings.getInstanceId()` | New function then needed for the cases if python needs the id byself to store something about for him himself or for the future works.
| `addon.setInstanceState(id, enabled)` | To allow on already present ones the enable / disable
| `addon.setInstanceName(id, name)` | To allow on already present ones the name change
| `state = settings.getInstanceState()` and<br>`state = addon.getInstanceState(id)` | For the sake of form, to have the counterpart to the “set”.
| `name = settings.getInstanceName()` and<br>`name = addon.getInstanceName(id)` | For the sake of form, to have the counterpart to the “set”.

About this it seems then also more hard to understand what does somewhere what.

phunkyfish commented:
> **What this change does**
>
> This allows a python script to edit add-on instance settings to be used within PVR add-ons.
> 
> To prevent conflicts from the user creating instances, add-on instances created by python start with id >1000.
>
> **The experience for the Python Dev**
>
> ```python
> import xbmcaddon
> import xbmcgui
>  
> addon       = xbmcaddon.Addon('pvr.demo')
> addonname   = addon.getAddonInfo('name')
> 
> newid = addon.getFreeNewInstanceId()
> addonsettingnew = addon.getSettings(newid)
> addonsettingnew.setString('host', '10.144.12.13')
> addon.setInstanceIdState(newid, True, 'By Python added')
> 
> # addon.deleteInstanceId(1001);
> 
> printstring = "Support Instances: " + str(addon.supportsInstanceSettings()) + "\n" \
>               "Used Instances:\n"
> for x in addon.getKnownInstanceIds():
>     printstring += "ID: " + str(x) + "\n"
>     printstring += "Test setting: " + addon.getSettings(x).getString('host') + "\n"
> 
> xbmcgui.Dialog().textviewer(addonname, printstring)
> ```
> 
> If the instance already present is it enough to call:
> 
> ```python
> id = 1001
> instancesetting = addon.getSettings(id)
> instancesetting.setString('host', '10.144.12.13')
> ```
> 
> By this the add-on becomes the `ADDON_STATUS SetInstanceSetting(const std::string& settingName, const kodi::addon::CSettingValue& settingValue)` call.

@enen92 based on the above example code can you explain how it should be in order to be able to be able to get this code in a merge state.

Maybe even re-write the code snippet and then we can see if that is possible, or if we need to make trade-offs somewhere. Thanks!
